lib/testing/zeroLowerTriangle.cpp

```diff
@@ -0,0 +1,100 @@
+
+#include "Config.hpp"          // for Config
+#include "StageFactory.hpp"    // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "buffer.h"            // for Buffer, mark_frame_empty, mark_frame_full, pass_metadata
+#include "bufferContainer.hpp" // for bufferContainer
+#include "kotekanLogging.hpp"  // for INFO
+
+#include <atomic>     // for atomic_bool
+#include <cstdint>    // for int32_t
+#include <exception>  // for exception
+#include <functional> // for _Bind_helper<>::type, bind, function
+#include <regex>      // for match_results<>::_Base_type
+#include <vector>     // for vector
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+REGISTER_KOTEKAN_STAGE(zeroLowerTriangle);
+
+zeroLowerTriangle::zeroLowerTriangle(Config& config, const std::string& unique_name,
+                                     bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&zeroLowerTriangle::main_thread, this)) {
+
+    // Apply config.
+    _num_elements = config.get<int32_t>(unique_name, "num_elements");
+    _num_local_freq = config.get<int32_t>(unique_name, "num_local_freq");
+    _samples_per_data_set = config.get<int32_t>(unique_name, "samples_per_data_set");
+    _sub_integration_ntime = config.get<int>(unique_name, "sub_integration_ntime");
+
+    input_buf = get_buffer("corr_in_buf");
+    register_consumer(input_buf, unique_name.c_str());
+    output_buf = get_buffer("corr_out_buf");
+    register_producer(output_buf, unique_name.c_str());
+}
+
+zeroLowerTriangle::~zeroLowerTriangle() {}
+
+void zeroLowerTriangle::main_thread() {
+
+    int input_frame_id = 0;
+    int output_frame_id = 0;
+
+    while (!stop_thread) {
+        int* input = (int*)wait_for_full_frame(input_buf, unique_name.c_str(), input_frame_id);
+        if (input == nullptr)
+            break;
+        int* output = (int*)wait_for_empty_frame(output_buf, unique_name.c_str(), output_frame_id);
+        if (output == nullptr)
+            break;
+
+        INFO("Zeroing out lower triangle for {:s}[{:d}] putting result in {:s}[{:d}]",
+             input_buf->buffer_name, input_frame_id, output_buf->buffer_name, output_frame_id);
+
+        int nt_inner = _sub_integration_ntime;
+        int n_outer = _samples_per_data_set / nt_inner;
+
+        for (int tout = 0; tout < n_outer; ++tout) {
+            for (int f = 0; f < _num_local_freq; ++f) {
+                for (int x = 0; x < _num_elements; ++x) {
+                    for (int y = 0; y < _num_elements; ++y) {
+                        int real = 0;
+                        int imag = 0;
+                        if (x <= y) {
+                            real = input[(((tout * _num_local_freq + f) * _num_elements + x)
+                                              * _num_elements
+                                          + y)
+                                             * 2
+                                         + 0];
```

Comment:
I don't love how clang-format handled this, it's quite hard to read...  We might want layout these by hand, and then include the clang-format ignore comment `// clang-format off` and `// clang-format on`